airflow/models/taskinstance.py

```diff
@@ -2821,15 +2821,16 @@ def tg2(inp):
 
     def clear_db_references(self, session):
         """
-        Clear DB references to XCom, TaskFail and RenderedTaskInstanceFields.
+        Clear DB references to XCom, TaskFail, TaskInstanceNote,
+        TaskReschedule and RenderedTaskInstanceFields.
```

Comment:
```suggestion
        Clear db tables what have a reference to this instance.
```

I wonder instead of hard-coding the tables if there’s something in SQLAlchemy that can introspect this.

Reply:
There is a function from sqlalchemy_utils that seems to do this but I couldn't get this working inside shell. We still need to filter out only foreign key references involving a combination of map_index to clear only those table's entries. sqlalchemy_utils is not a dependency so we need to copy that function I guess.

https://sqlalchemy-utils.readthedocs.io/en/latest/foreign_key_helpers.html#get-referencing-foreign-keys

Reply:
Postgres gives the referenced by output

```
airflow_user@127:airflow_db> \d task_instance;
+----------------------+-----------------------------+---------------------------------+
| Column               | Type                        | Modifiers                       |
|----------------------+-----------------------------+---------------------------------|
| task_id              | character varying(250)      |  not null                       |
| dag_id               | character varying(250)      |  not null                       |
| run_id               | character varying(250)      |  not null                       |
| map_index            | integer                     |  not null default '-1'::integer |
| start_date           | timestamp with time zone    |                                 |
| end_date             | timestamp with time zone    |                                 |
| duration             | double precision            |                                 |
| state                | character varying(20)       |                                 |
| try_number           | integer                     |                                 |
| max_tries            | integer                     |  default '-1'::integer          |
| hostname             | character varying(1000)     |                                 |
| unixname             | character varying(1000)     |                                 |
| job_id               | integer                     |                                 |
| pool                 | character varying(256)      |  not null                       |
| pool_slots           | integer                     |  not null                       |
| queue                | character varying(256)      |                                 |
| priority_weight      | integer                     |                                 |
| operator             | character varying(1000)     |                                 |
| queued_dttm          | timestamp with time zone    |                                 |
| queued_by_job_id     | integer                     |                                 |
| pid                  | integer                     |                                 |
| executor_config      | bytea                       |                                 |
| updated_at           | timestamp with time zone    |                                 |
| external_executor_id | character varying(250)      |                                 |
| trigger_id           | integer                     |                                 |
| trigger_timeout      | timestamp without time zone |                                 |
| next_method          | character varying(1000)     |                                 |
| next_kwargs          | json                        |                                 |
+----------------------+-----------------------------+---------------------------------+
Indexes:
    "task_instance_pkey" PRIMARY KEY, btree (dag_id, task_id, run_id, map_index)
    "ti_dag_run" btree (dag_id, run_id)
    "ti_dag_state" btree (dag_id, state)
    "ti_job_id" btree (job_id)
    "ti_pool" btree (pool, state, priority_weight)
    "ti_state" btree (state)
    "ti_state_incl_start_date" btree (dag_id, task_id, state) INCLUDE (start_date)
    "ti_state_lkp" btree (dag_id, task_id, run_id, state)
    "ti_trigger_id" btree (trigger_id)
Foreign-key constraints:
    "task_instance_dag_run_fkey" FOREIGN KEY (dag_id, run_id) REFERENCES dag_run(dag_id, run_id) ON DELETE CASCADE
    "task_instance_trigger_id_fkey" FOREIGN KEY (trigger_id) REFERENCES trigger(id) ON DELETE CASCADE
Referenced by:
    TABLE "task_fail" CONSTRAINT "task_fail_ti_fkey" FOREIGN KEY (dag_id, task_id, run_id, map_index) REFERENCES task_instance(dag_id, task_id, run_id, map_index) ON DELETE CASCADE
    TABLE "task_map" CONSTRAINT "task_map_task_instance_fkey" FOREIGN KEY (dag_id, task_id, run_id, map_index) REFERENCES task_instance(dag_id, task_id, run_id, map_index) ON DELETE CASCADE
    TABLE "task_reschedule" CONSTRAINT "task_reschedule_ti_fkey" FOREIGN KEY (dag_id, task_id, run_id, map_index) REFERENCES task_instance(dag_id, task_id, run_id, map_index) ON DELETE CASCADE
    TABLE "xcom" CONSTRAINT "xcom_task_instance_fkey" FOREIGN KEY (dag_id, task_id, run_id, map_index) REFERENCES task_instance(dag_id, task_id, run_id, map_index) ON DELETE CASCADE
    TABLE "task_instance_note" CONSTRAINT "task_instance_note_ti_fkey" FOREIGN KEY (dag_id, task_id, run_id, map_index) REFERENCES task_instance(dag_id, task_id, run_id, map_index) ON DELETE CASCADE
    TABLE "rendered_task_instance_fields" CONSTRAINT "rtif_ti_fkey" FOREIGN KEY (dag_id, task_id, run_id, map_index) REFERENCES task_instance(dag_id, task_id, run_id, map_index) ON DELETE CASCADE

Time: 0.025s
```

Reply:
A rough attempt in Python. This needs all models to be imported so that they are registered.

```python
import pprint
import time

from airflow.models.taskinstance import *
from airflow.models.dag import *
from airflow.models.trigger import *
from airflow.www.fab_security.sqla.models import *


def get_references(model):
    references = {}

    for table in model.__table__.metadata.tables.values():
        foreign_key_constraints = table.foreign_key_constraints
        for constraint in foreign_key_constraints:
            refers = all(
                element.references(model.__table__) for element in constraint.elements
            )
            if refers:
                references[table.name] = {
                    "columns": [element.column.name for element in constraint.elements],
                    "constraint": constraint.name,
                }

    return references


start = time.time()
pprint.pp(get_references(TaskInstance))
duration = time.time() - start
print(f"Took {duration} seconds")
```

```
time python /tmp/foo.py
{'task_fail': {'columns': ['dag_id', 'task_id', 'run_id', 'map_index'],
               'constraint': 'task_fail_ti_fkey'},
 'task_map': {'columns': ['dag_id', 'task_id', 'run_id', 'map_index'],
              'constraint': 'task_map_task_instance_fkey'},
 'task_reschedule': {'columns': ['dag_id', 'task_id', 'run_id', 'map_index'],
                     'constraint': 'task_reschedule_ti_fkey'},
 'xcom': {'columns': ['dag_id', 'task_id', 'run_id', 'map_index'],
          'constraint': 'xcom_task_instance_fkey'},
 'task_instance_note': {'columns': ['dag_id', 'task_id', 'run_id', 'map_index'],
                        'constraint': 'task_instance_note_ti_fkey'}}
Took 0.001352071762084961 seconds
```